### PR TITLE
Test to verify setting osd_memory_target on osd/host level

### PR DESCRIPTION
**Pull request by msainiRedhat**
# Description

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

**Comment by vamahaja** [comment]
Testruns -

    http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-yam5z
    http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-8v5vg

File: cli/ceph/ceph.py [hunk 1]
```diff
@@ -84,3 +77,9 @@ def osd(self, command, **kw):
         cmd = f"{self.base_cmd} osd {command}"
         cmd += build_cmd_from_args(**kw)
         return self.execute(sudo=True, check_ec=False, long_running=False, cmd=cmd)
+
+    def ceph_config(self, command, **kw):
+        """Executes ceph config specific commands"""
+        cmd = f"{self.base_cmd} config {command}"
+        cmd += build_cmd_from_args(**kw)
```
Comment (vamahaja) [inline cli/ceph/ceph.py:84]: Why separate statement ? Can't it be merged while declaring cmd ?

File: cli/ceph/config.py [hunk 2]
```diff
@@ -17,3 +17,28 @@ def set(self, key, value):
         """
         cmd = f"{self.base_cmd} set {key} {value}"
         return self.execute(sudo=True, cmd=cmd)
+
+    def set_osd_memory_autotune(self, value):
```
Comment (vamahaja) [inline cli/ceph/config.py:21]: `set_osd_memory_autotune` and `set_host_osd_memory_target` can be merged with `set` method

File: cli/ceph/config.py [hunk 3]
```diff
@@ -17,3 +17,28 @@ def set(self, key, value):
...
+    def set_host_osd_memory_target(self, host, value):
+        """set osd memory target value on Host
+        Args:
+            host (str) : Host on which value needs to be set
+            value (int) : osd memory target value
+        """
+        cmd = f"{self.base_cmd} set osd/host:{host} osd_memory_target {value}"
+        return self.execute(sudo=True, cmd=cmd)
+
+    def get_osd_memory_target(self, osd_id):
```
Comment (vamahaja) [inline cli/ceph/config.py:38]: this should be converted to generic `get` method

File: tests/cephadm/test_osd_memory_target.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    out = cephadm.ceph.ceph_config("dump", **kw)
+    data = json.loads(out[0])
+    value_found = False
+    for data_config in data:
+        if data_config["name"] == "osd_memory_target" and data_config["value"] == str(
+            value
+        ):
+            value_found = True
+            break
+    if value_found:
```
Comment (vamahaja) [inline tests/cephadm/test_osd_memory_target.py:47]: More readable way -
```
    if not found:
        raise ConfigError(
            "osd_memory_target not found or has an incorrect value in ceph config"
        )
    log.info("osd memory target is validated in ceph config")

```

File: tests/cephadm/test_osd_memory_target.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    # Verify that the option is set in ceph config dump
+    kw = {"format": "json-pretty"}
+    out = cephadm.ceph.ceph_config("dump", **kw)
+    data = json.loads(out[0])
+    value_found = False
+    for data_config in data:
+        if data_config["name"] == "osd_memory_target" and data_config["value"] == str(
+            value
+        ):
+            value_found = True
```
Comment (vamahaja) [inline tests/cephadm/test_osd_memory_target.py:45]: change this to `found`

File: tests/cephadm/test_osd_memory_target.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            break
+    if value_found:
+        log.info("osd memory target is validated in ceph config")
+    else:
+        raise ConfigError(
+            "osd_memory_target not found or has an incorrect value in ceph config"
+        )
+
+    # Verify osd memory target for osd on the host matches the host level value
+    out = cephadm.ceph.osd("tree", **kw)
```
Comment (vamahaja) [inline tests/cephadm/test_osd_memory_target.py:55]: instead add method `tree` here - https://github.com/red-hat-storage/cephci/blob/master/cli/ceph/osd/osd.py

File: tests/cephadm/test_osd_memory_target.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    data = json.loads(out[0])
+    for item in data["nodes"]:
+        if item["type"] == "host":
+            if item["name"] == hostname:
+                osd_list = item["children"]
+                osd_id = "osd." + str(osd_list[0])
```
Comment (vamahaja) [inline tests/cephadm/test_osd_memory_target.py:61]: Change this to `f"osd.{str(osd_list[0])}"`

**Review by vamahaja — CHANGES_REQUESTED** [review]

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: cli/ceph/ceph.py [hunk 8]
```diff
@@ -84,3 +84,8 @@ def osd(self, command, **kw):
         cmd = f"{self.base_cmd} osd {command}"
         cmd += build_cmd_from_args(**kw)
         return self.execute(sudo=True, check_ec=False, long_running=False, cmd=cmd)
+
+    def ceph_config(self, command, **kw):
```
Comment (vamahaja) [inline cli/ceph/ceph.py:88]: Move this method to [1] accordingly
https://github.com/red-hat-storage/cephci/blob/master/cli/ceph/config.py

File: cli/ceph/config.py [hunk 9]
```diff
@@ -9,11 +9,24 @@ def __init__(self, nodes, base_cmd):
...
+            daemon (str) : daemon with or without path
         """
-        cmd = f"{self.base_cmd} set {key} {value}"
+        if daemon is None:
+            cmd = f"{self.base_cmd} set {key} {value}"
+        else:
+            cmd = f"{self.base_cmd} set {daemon} {key} {value}"
+        return self.execute(sudo=True, cmd=cmd)
+
+    def get(self, key, value):
```
[the remark this answers is not in the document]
Comment (vamahaja) [reply cli/ceph/config.py:25]: `def get(self, who, key=None):`

File: cli/ceph/config.py [hunk 10]
```diff
@@ -9,11 +9,24 @@ def __init__(self, nodes, base_cmd):
...
+        """Fetch the config value
+        Args:
+            key (str) :  config key
+            value (str) : config key value
+        """
+        cmd = f"{self.base_cmd} get {key} {value}"
```
Comment (vamahaja) [inline cli/ceph/config.py:31]:
```
cmd = f"{self.base_cmd} get {who}"

if key:
     cmd += "{key}"
```

**Comment by msainiRedhat** [comment]
Logs - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-98yiv/

**Review by vamahaja — APPROVED** [review]

**Pull request merged**